src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs

```diff
@@ -16,15 +16,34 @@ public LocationRepository(StatisticsDbContext context) : base(context)
 
         public Dictionary<GeographicLevel, IEnumerable<IObservationalUnit>> GetObservationalUnits(Guid subjectId)
         {
-            var locations = GetLocationsGroupedByGeographicLevel(subjectId);
-            return GetObservationalUnits(locations);
+            var observations = _context.Observation
```

Comment:
In data files where there are, say, 250,000 rows, we want to avoid pulling out 250,000 records out of the database which could take some time for EF and some chunk of memory usage too.

Could you just confirm that the 2 usages of this query in the `GetObservationalUnits(IQueryable<Observation> observations)` method don't cause this list to be pulled back early before initiating both their subqueries?  I'm guessing they're fine to be honest:

```
var geographicLevels = observations
                .Select(o => o.GeographicLevel)
                .Distinct()
                .ToList();
```

and

```
var locations = observations
                            .Include(observation => observation.Location)
                            .Where(observation => observation.GeographicLevel == geographicLevel)
                            .Select(observation => observation.Location)
                            .Distinct()
                            .ToList();
```

I think that both of these will just grab what they need without triggering the full retrieval of 250,000 Observation instances.

Reply:
Discussed this with Duncan on Slack. The original code is written as it is to avoid multiple calls to the Observation table. So I've changed the code to do this, but have submitted it in a separate PR: https://github.com/dfe-analytical-services/explore-education-statistics/pull/2797